### ISSUE TITLE
user-defined triton kernels + inductor stride re-ordering can lead to silent incorrectness

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#130243]
Inductor is allowed to reorder the strides of intermediate tensors. user-defined triton kernels have different semantics depending on the striding of their input tensors. Nothing is stopping inductor from reordering the strides of inputs to triton kernels, so this can lead to silent incorrectness:

```
import torch
import triton
import triton.language as tl

@triton.jit
def add_kernel(
    in_ptr0,
    in_ptr1,
    out_ptr,
    n_elements,
    BLOCK_SIZE: "tl.constexpr",
):
    pid = tl.program_id(axis=0)
    block_start = pid * BLOCK_SIZE
    offsets = block_start + tl.arange(0, BLOCK_SIZE)
    mask = offsets < n_elements
    x = tl.load(in_ptr0 + offsets, mask=mask)
    y = tl.load(in_ptr1 + offsets, mask=mask)
    output = x + y
    tl.store(out_ptr + offsets, output, mask=mask)

x = torch.randn(2, 2, device="cuda")
other = torch.randn(2, 2, device="cuda")

def f(x, other):
    y = x.t().contiguous().t()
    z = y.sin().t()
    grid = (z.numel(),)
    out = torch.empty_like(other)
    add_kernel[grid](z, other, out, z.numel(), BLOCK_SIZE=16)
    return out

f_compile = torch.compile(f)

out = f(x, other)
out_compile = f_compile(x, other)
print(out)
print(out_compile)
assert torch.allclose(out_compile, out)
```

NB: we have a similar problem with custom operators that we're working on fixing (https://github.com/pytorch/pytorch/issues/124647). Maybe we can share some of the infra there.

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @muchulee8 @ColinPeppler @amjames @desertfire @oulgen @aakhundov @davidberard98 @anijain2305 @peterbell10

**Comment by ezyang** [comment]
This just sounds like, by default we should force stride to be preserved into user defined triton kernels...

**Comment by oulgen** [comment]
@ezyang : @zou3519 and I chatted offline about this, the plan is to build something like `constrain_fx_strides` and apply it to all tensor inputs of triton kernels, as well as custom ops.

**Comment by zou3519** [comment]
To clarify, constrain_to_fx_strides exists (https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L2026), it has some issues that need to be fixed:
1) it leads to additional copies at times
2) it has some silent incorrectness when it comes to constraining the FX strides of an input that is being mutated

**Comment by blaine-rister** [comment]
> To clarify, constrain_to_fx_strides exists (https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L2026), it has some issues that need to be fixed before we use it by default:
> 
> 1. it leads to additional copies at times
> 2. it has some silent incorrectness when it comes to constraining the FX strides of an input that is being mutated

@ezyang does `constrain_to_fx_strides` constrain the exact strides to be used? Or only their order? I'm wondering if the constraint is strong enough to support things like discontiguous strides.

**Comment by zou3519** [comment]
> @zou3519 does constrain_to_fx_strides constrain the exact strides to be used? Or only their order? I'm wondering if the constraint is strong enough to support things like discontiguous strides.

It constrains to the order, not the exact strides. This should be sufficient for the majority of use cases (that have things like channels-last tensors or row-major matrices). Do you have a use case where the actual strides matter?

**Comment by blaine-rister** [comment]
> > @zou3519 does constrain_to_fx_strides constrain the exact strides to be used? Or only their order? I'm wondering if the constraint is strong enough to support things like discontiguous strides.
> 
> It constrains to the order, not the exact strides. This should be sufficient for the majority of use cases (that have things like channels-last tensors or row-major matrices). Do you have a use case where the actual strides matter?

I can think of a few use cases:
  - Writing a kernel that can take a view as input.
  - Writing kernels for hardware that imposes memory alignment constraints or otherwise likes to see padded data. (The use case which motivated #130394.)
  
Granted this is probably niche and won't affect most users, but triton does not require inputs to be contiguous. We have some tests for discontiguous inputs [here](https://github.com/pytorch/pytorch/blob/main/test/inductor/test_torchinductor_strided_blocks.py). So if we really want to put this issue to rest, we might consider enforcing exact strides as opposed to just the ordering.

**Comment by zou3519** [comment]
Raising priority for awareness

**Comment by FindHao** [comment]
I'm fixing a similar issue here. https://github.com/pytorch/pytorch/pull/130956 But I found the problem in this issue is not exact same with it. In this case, it doesn't go through the [`if (is_output` branch](https://github.com/pytorch/pytorch/pull/130956/files#diff-e7f291bf7d847fea77a46bd4c0ce6ee0eacf851714a397b4271e54f36fedf369R1233-R1243). But I believe some definitions can be shared, e.g., `freeze_layout_with_actual_strides`.

**Comment by daewook** [comment]
while waiting for this fix, is there currently any workaround to force the stride to be preserved before the user-defined triton kernel?

**Comment by zou3519** [comment]
Need an opinion (@oulgen @aakhundov + anyone else on this thread): we have three options:
1. by default, inductor will respect the dense stride order of the inputs to triton kernels (for example, if the inputs are contiguous, row-major, or column-major, inductor won't change that. However, if the strides are non-dense like [5], inductor will change it to be dense [1]).
2. by default, inductor will respect exact strides to the input to triton kernels
3. [the current status] by default, inductor will not respect the strides and can change them to be whatever

Which option do we think is a good default? For custom ops, we're targeting (1), because custom ops are commonly written to accept contiguous or dense inputs. But I haven't written or seen enough triton kernels to say.

Independently of this default, we will offer a way for users to override the default.

**Comment by ezyang** [comment]
@zou3519 I have a spicy take... what if we just error by default if an input into a custom op is non-contiguous and we don't know if (3) is ok

**Comment by daewook** [comment]
Let me share my usage for a datapoint. I'm using a Triton implementation for a custom autograd function (both forward and backward). This is fairly common practice. As noted in [this issue](https://github.com/pytorch/pytorch/issues/115344), "tensor.stride is banned in autograd.function."

Assume my Triton backward function receives dy (gradient of output y) as input. I can't pass the strides of dy to the function due to above restriction. Instead, I call dy.contiguous() before calling the Triton function and assume dy is contiguous inside the Triton backward code.

In this case, it seems option (2) is the only viable solution unless I can call .stride for all tensor inputs of the autograd backward function.

**Comment by blaine-rister** [comment]
Option (2) seems the safest to me, but I'm biased towards a somewhat atypical use case. I'm working with a hardware accelerator which imposes fairly strict memory alignment requirements, and so it often likes to see discontiguous tensors. Since Triton works with pointers, our kernels have to be aware of this nonstandard memory layout, and might not work properly for other layouts. 

It's useful to be able to tell Inductor that we want to pass discontiguous tensors to the kernels. I'm not sure that should be the default, but it would be nice if it's at least an option.

**Comment by jansel** [comment]
We should force inductor match eager strides for all custom ops (with perhaps an op tag to disable this).

There is also https://github.com/pytorch/pytorch/blob/a4b44dd2ef99f56ea5a517468ad8497740955064/torch/_inductor/inductor_prims.py#L81-L85

**Comment by daewook** [comment]
Hi, since this is still not fixed, is there any way to work around this on the user side? My current solution is to add @torch.compiler.disable around the code that uses user-defined triton kernel but it's slow.
